components/score_tracker.py

```diff
@@ -0,0 +1,155 @@
+
+    def __init__(self) -> None:
+        # confidences for all pieces on the field, including other side
+        self.confidences: np.ndarray = np.zeros(3 * 3 * 3 * 2)
+
+    def add_vision(self, node_location: NodeLocation, node_value: bool) -> None:
+        """
+        Modify the confidence of the piece in location for node_id
+        node_id: The id of the node
+        node_value: Weather there is a piece there
+        """
+        idx = node_location.get_index()
+        cur_confidence = self.confidences[idx]
+        if node_value and cur_confidence < 1:
+            self.confidences[idx] += (
+                self.SIGHTING_TRUE_WEIGHT * self.control_loop_wait_time
+            )
+        if not node_value:
+            self.confidences[idx] += (
+                self.SIGHTING_FALSE_WEIGHT * self.control_loop_wait_time
+            )
+
+    def get_row(self, row: int) -> npt.NDArray[np.bool_]:
+        return self.confidences[row * 9 : (row + 1) * 9] > self.CONFIDENCE_THRESHOLD
+
+    def add_piece(self, node_location: NodeLocation) -> None:
+        """
+        A function to add a piece as 100% on position of layer of node_id
+        node_id: The node (1-8)
+        layer: The layer (0-2)
+        position: The postion (0-2)
+        """
+        self.confidences[node_location.get_index()] = self.PLACE_WEIGHT
+
+    def evaluate_place_location(
+        self, location: NodeLocation, piece_type: GamePiceType
+    ) -> float:
+        """
+        Give a reletive score of how good it is to score in this position,
+        roughly equivilant to the score after placing.
+        location: Where to place the piece
+        piece_type: What piece
+        """
+        if not location.can_place_piece(piece_type):
+            return -1
+
+        # get turn confidences into bools
+        row = self.get_row(location.row)
+        row[location.column] = True
+        return evaluate_row(row, location.row)
+
+    def get_node_picklist(self, piece_type: GamePiceType) -> list[NodeLocation]:
+        """Gets a sorted list of which locations to put piece_type"""
+        # capture the piece type in a function to sort with
+        def evaluate_func(location):
+            return self.evaluate_place_location(location, piece_type)
+
+        all_locations: list[NodeLocation] = []
+        for side in (FieldSide.BLUE, FieldSide.RED):
+            for row in range(2):
+                for col in range(9):
+                    all_locations.append(NodeLocation(col, row, side))
+
+        return sorted(all_locations, key=evaluate_func, reverse=True)
+
+
+def evaluate_row(row_values: npt.NDArray[np.bool_], row: int) -> float:
+    """
+    Evaluate how good a state of a row is
+    row_values: bool array or if there are pieces in nodes
+    row: which row (0 bottom, 1 middle, 2 top)
+    """
+    total_score = 0.0
+    piece_value = 0
+    if row == 0:
+        piece_value = 2
+    if row == 1:
+        piece_value = 3
+    if row == 2:
```

Comment:
Pick one:
- use `elif`
- use a mapping from row to point value

Reply:
what if the row is in superposition between 1 and 2????

Reply:
We've observed it so the state will have collapsed by the time we reach this function